riscv/src/runtime.rs

```diff
@@ -90,51 +89,88 @@ impl Runtime {
+                    link ~> val2_col = regs.mload(Y, STEP + 1)
+                    link ~> val3_col = binary.and(val1_col, val2_col + Z)
+                    link ~> regs.mstore(W, STEP + 3, val3_col);"#,
+                r#"instr or X, Y, Z, W
+                    link ~> val1_col = regs.mload(X, STEP)
+                    link ~> val2_col = regs.mload(Y, STEP + 1)
+                    link ~> val3_col = binary.or(val1_col, val2_col + Z)
+                    link ~> regs.mstore(W, STEP + 3, val3_col);"#,
+                r#"instr xor X, Y, Z, W
+                    link ~> val1_col = regs.mload(X, STEP)
+                    link ~> val2_col = regs.mload(Y, STEP + 1)
+                    link ~> val3_col = binary.xor(val1_col, val2_col + Z)
+                    link ~> regs.mstore(W, STEP + 3, val3_col);"#,
             ],
             0,
-            ["x10 <== and(x10, x10);"],
+            ["and 0, 0, 0, 0;"],
         );
 
         r.add_submachine(
             "std::machines::shift::Shift",
             None,
             "shift",
             [
-                "instr shl Y, Z -> X link ~> X = shift.shl(Y, Z);",
-                "instr shr Y, Z -> X link ~> X = shift.shr(Y, Z);",
+                r#"instr shl X, Y, Z, W
+                    link ~> val1_col = regs.mload(X, STEP)
+                    link ~> val2_col = regs.mload(Y, STEP + 1)
+                    link ~> val3_col = shift.shl(val1_col, val2_col + Z)
+                    link ~> regs.mstore(W, STEP + 3, val3_col);"#,
+                r#"instr shr X, Y, Z, W
+                    link ~> val1_col = regs.mload(X, STEP)
+                    link ~> val2_col = regs.mload(Y, STEP + 1)
+                    link ~> val3_col = shift.shr(val1_col, val2_col + Z)
+                    link ~> regs.mstore(W, STEP + 3, val3_col);"#,
             ],
             0,
-            ["x10 <== shl(x10, x10);"],
+            ["shl 0, 0, 0, 0;"],
         );
 
         r.add_submachine(
             "std::machines::split::split_gl::SplitGL",
             None,
             "split_gl",
-            ["instr split_gl Z -> X, Y link ~> (X, Y) = split_gl.split(Z);"],
+            [r#"instr split_gl X, Z, W
+                    link ~> val1_col = regs.mload(X, STEP)
+                    link ~> (val3_col, val4_col) = split_gl.split(val1_col)
+                    link ~> regs.mstore(Z, STEP + 2, val3_col)
+                    link ~> regs.mstore(W, STEP + 3, val4_col);"#],
             0,
-            ["x10, x11 <== split_gl(x10);", "x10 <=X= 0;", "x11 <=X= 0;"],
+            ["split_gl 0, 0, 0;"],
         );
 
         // Base syscalls
         r.add_syscall(
             Syscall::Input,
-            ["x10 <=X= ${ std::prover::Query::Input(std::convert::int(std::prover::eval(x10))) };"],
+            [
+                "val1 <== get_reg(10);",
+                "val1 <=X= ${ std::prover::Query::Input(std::convert::int(std::prover::eval(val1))) };",
+                "set_reg 10, val1;"
```

Comment:
```suggestion
                "set_reg 10, ${ std::prover::Query::Input(std::convert::int(std::prover::eval(val1))) };"
```
Wouldn't this work?

Reply:
BTW, even with this change, it is pretty silly that we need to:
- add a register (`val1`), and
- run an instruction

just to get the value to pass to the query! If here was a way to access the register memory from a PIL function, we could get rid of this. I don't see an easy way to do that though, but we should keep this in the back of our head. We do run the input syscall quite often when receiving larger amounts of data, don't we?

Maybe it's worth adding a comment?

Reply:
yea I also don't know how to solve this atm =/ but agree we should focus on it soon